jl4-lsp/src/LSP/L4/Viz/Ladder.hs

```diff
@@ -30,67 +49,108 @@ isDevMode = False
 -- Monad
 ------------------------------------------------------
 
-newtype Viz a = MkViz {getVizE :: VizState -> (Either VizError a, VizState)}
+newtype Viz a = MkViz {getViz :: VizEnv -> VizState -> (Either VizError a, VizState)}
   deriving
-    (Functor, Applicative, Monad, MonadState VizState, MonadError VizError)
-    via ExceptT VizError (State VizState)
+    (Functor, Applicative, Monad, MonadState VizState, MonadError VizError, MonadReader VizEnv)
+    via ReaderT VizEnv (ExceptT VizError (State VizState))
+
+-- | A 'local' env
+newtype VizEnv = MkVizEnv { localDecls :: [LocalDecl Resolved]}
 
-mkVizEnv :: LSP.VersionedTextDocumentIdentifier -> TC.Substitution -> Bool -> VizEnv
-mkVizEnv verTxtDocId substitution shouldSimplify =
+mkVizConfig :: LSP.VersionedTextDocumentIdentifier -> TC.Substitution -> Bool -> VizConfig
+mkVizConfig verTxtDocId substitution shouldSimplify =
   let moduleUri = toNormalizedUri verTxtDocId._uri
-  in MkVizEnv
+  in MkVizConfig
     { moduleUri
     , verTxtDocId
     , substitution
     , shouldSimplify
     }
 
-data VizEnv = MkVizEnv
+data VizConfig = MkVizConfig
   { moduleUri      :: !NormalizedUri
   , verTxtDocId    :: !LSP.VersionedTextDocumentIdentifier
   , substitution   :: !TC.Substitution  -- might be more futureproof to use TypeCheckResult
   , shouldSimplify :: !Bool             -- ^ whether to simplify the expression
   }
   deriving stock (Show, Generic, Eq)
 
-data VizState =
-  MkVizState
-    { env :: !VizEnv
-    , maxId :: !ID
-    }
-  deriving stock (Show, Generic, Eq)
 
-mkInitialVizState :: VizEnv -> VizState
-mkInitialVizState env =
+data VizState = MkVizState
```

Comment:
I might be pre-maturely optimizing with the bang patterns --- haven't done any profiling